cmd/nerdctl/cp.go

```diff
@@ -0,0 +1,609 @@
+		cleanedPath := filepath.Clean(trimmedPath)
+		prefix := copyutil.StripPathShortcuts(cleanedPath)
+
+		// header.Name is a name of the REMOTE file, so we need to create
+		// a remotePath that will goes through appropriate cleaning process
+		destFileName := path.Join(dstPath, filepath.Clean(header.Name[len(prefix):]))
+
+		if !copyutil.IsRelative(dstPath, destFileName) {
+			logrus.Warnf("skipping %q: file is outside target destination", destFileName)
+			continue
+		}
+
+		// docker does not manage the creation of non-existent path
+		if err := os.MkdirAll(filepath.Dir(destFileName), 0755); err != nil {
+			return err
+		}
+		if header.FileInfo().IsDir() {
+			if err := os.MkdirAll(destFileName, 0755); err != nil {
+				return err
+			}
+			continue
+		}
+
+		if dstPath == "-" {
+			_, err = io.Copy(cmd.OutOrStdout(), tarReader)
+			return err
+		}
+
+		isquiet, err := cmd.Flags().GetBool("quiet")
+		if err != nil {
+			return err
+		}
+
+		outFile, err := os.Create(destFileName)
+		if err != nil {
+			return err
+		}
+		defer func() {
+			if err := outFile.Close(); err != nil {
+				logrus.WithError(err).Warnf("failed to close: %q", outFile.Name())
+			}
+		}()
+
+		var writer io.Writer
+		writer = outFile
+
+		if !isquiet {
+			bar := progressbar.DefaultBytes(
+				header.Size,
+				"downloading",
+			)
+			writer = io.MultiWriter(outFile, bar)
+		}
+
+		if _, err := io.Copy(writer, tarReader); err != nil {
+			return err
+		}
+
+	}
+
+	return nil
+}
+
+func (t *TarPipe) startReadFrom(n uint64) {
+	tarBinary, err := copyutil.TarBinary()
+	if err != nil {
+		hint := fmt.Sprintf("`tar` binary needs to be installed")
+		logrus.Errorf(hint+": %w", err)
+	}
+
+	var tarArgs []string
+	tarArgs = []string{"-cf", "-", "-C", filepath.Join("/proc/", fmt.Sprint(t.option.containerPid), "/root"), strings.TrimLeft(t.src, `/\`)}
+
+	command := exec.Command(tarBinary, tarArgs...)
+	command.Stdout = t.outStream
+
+	go func() {
+		defer t.outStream.Close()
+		err = command.Run()
+		if err != nil {
+			logrus.Errorf("failed to exec: %v", err)
+		}
+	}()
+}
+
+func (t *TarPipe) Read(p []byte) (n int, err error) {
+	n, err = t.reader.Read(p)
+	if err != nil {
+		return 0, err
+	}
+	return n, nil
+}
+
+func initTarPipe(copyConfig *cpConfig) *TarPipe {
+	t := TarPipe{
+		src:      copyConfig.sourcePath,
+		dst:      copyConfig.destPath,
+		option:   copyConfig,
+		readChan: make(chan uint64), // value: byte number
+	}
+	t.reader, t.outStream = io.Pipe()
+
+	return &t
+}
+
+func cpShellComplete(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	if len(args) == 0 {
+		// show running container names
+		statusFilterFn := func(st containerd.ProcessStatus) bool {
+			return st == containerd.Running
+		}
+		return shellCompleteContainerNames(cmd, statusFilterFn)
+	} else {
+		return nil, cobra.ShellCompDirectiveNoFileComp
+	}
+}
```

Comment:
Doesn't seem to work as expected.

`nerdctl cp <TAB>` should show local filenames